File: src/main/java/formationJpa/entity/recette/Recette.java
package formationJpa.entity.recette;

import java.util.List;

import javax.persistence.Column;
import javax.persistence.DiscriminatorColumn;
import javax.persistence.DiscriminatorType;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.ForeignKey;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;

import formationJpa.entity.AssociationRecetteCommentaires;
import formationJpa.entity.Utilisateur;
import formationJpa.entity.Ingredients.AssociationIngredientsRecette;
import formationJpa.entity.Ingredients.Ingredient;
import formationJpa.entity.tag.AssociationTagRecette;
import formationJpa.entity.tag.Tag;

@Entity
@Table(name = "Recipe")
@SequenceGenerator(name = "seqRecipe", sequenceName = "seq_recipe", initialValue = 100, allocationSize = 1)
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "type", discriminatorType = DiscriminatorType.STRING, length = 1)
public  abstract class Recette {
	@Column(name = "name_recipe",length = 150)
	private String nom;
	@Id
	@GeneratedValue(strategy = GenerationType.SEQUENCE,generator = "seqRecipe")
	private Integer id;
	@Column(name = "nbperson_recipe",length = 150)
	private Integer nbPersonne;
	@Column(name = "duration_recipe",length = 150)
	private Integer temps;
	@OneToMany(mappedBy = "recette")
	private List<AssociationTagRecette> tags;
	@OneToMany(mappedBy = "recette")
	private List<AssociationIngredientsRecette> ingredients;
	@OneToMany(mappedBy ="id_recette")
	private List<EtapeRecette> etapes;
	@OneToMany(mappedBy ="commentaires")
	private List<AssociationRecetteCommentaires>  commentaires;
	@ManyToOne
	@JoinColumn(name = "auteur", foreignKey = @ForeignKey(name = "recette_auteur_fk"))
	private Utilisateur auteur;
	@Column(name = "cost", length = 2)
	@Enumerated(EnumType.STRING)
	private Couts cout;
	@Column(name = "difficulte", length = 2)
	@Enumerated(EnumType.STRING)
	private Difficulte difficulte;
	
	
	public Recette() {
		
	}

	public Recette(String nom, Integer id, Integer nbPersonne, Integer temps) {
		this.nom = nom;
		this.id = id;
		this.nbPersonne = nbPersonne;
		this.temps = temps;
	}

	public String getNom() {
		return nom;
	}

	public void setNom(String nom) {
		this.nom = nom;
	}

	public Integer getId() {
		return id;
	}

	public void setId(Integer id) {
		this.id = id;
	}

	public Integer getNbPersonne() {
		return nbPersonne;
	}

	public void setNbPersonne(Integer nbPersonne) {
		this.nbPersonne = nbPersonne;
	}

	public Integer getTemps() {
		return temps;
	}

	public void setTemps(Integer temps) {
		this.temps = temps;
	}

	public List<AssociationTagRecette> getTags() {
		return tags;
	}

	public void setTags(List<AssociationTagRecette> tags) {
		this.tags = tags;
	}

	public List<AssociationIngredientsRecette> getIngredients() {
		return ingredients;
	}

	public void setIngredients(List<AssociationIngredientsRecette> ingredients) {
		this.ingredients = ingredients;
	}

	public List<EtapeRecette> getEtapes() {
		return etapes;
	}

	public void setEtapes(List<EtapeRecette> etapes) {
		this.etapes = etapes;
	}

	public List<AssociationRecetteCommentaires> getCommentaires() {
		return commentaires;
	}

	public void setCommentaires(List<AssociationRecetteCommentaires> commentaires) {
		this.commentaires = commentaires;
	}

	public Utilisateur getAuteur() {
		return auteur;
	}

	public void setAuteur(Utilisateur auteur) {
		this.auteur = auteur;
	}

	public Couts getCout() {
		return cout;
	}

	public void setCout(Couts cout) {
		this.cout = cout;
	}

	public Difficulte getDifficulte() {
		return difficulte;
	}

	public void setDifficulte(Difficulte difficulte) {
		this.difficulte = difficulte;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((id == null) ? 0 : id.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Recette other = (Recette) obj;
		if (id == null) {
			if (other.id != null)
				return false;
		} else if (!id.equals(other.id))
			return false;
		return true;
	}
	
	
}

File: angular/src/app/model/recette.ts
import { Ingredient } from './ingredient';
import { EtapeRecette } from './etape-recette';
import { Commentaire } from './commentaire';
import { Cout } from './cout.enum';
import { Difficulte } from './difficulte.enum';
import { ImageModel } from './image-model';
import { Utilisateur } from '../user/model/utilisateur';

export class Recette {
  constructor(
    private _id?: number,
    private _nom?: string,
    private _type?: string,
    private _nbPersonne?: number,
    private _temps?: number,
    private _cout?: Cout,
    private _difficulte?: Difficulte,
    private _picture?: ImageModel,
    private _dateCreation?: Date,
    private _tags?: string[],
    private _ingredients?: Ingredient[],
    private _etapes?: EtapeRecette[],
    private _commentaires?: Commentaire[],
    private _auteur?: Utilisateur
  ) {}

  get id(): number {
    return this._id;
  }

  set id(value: number) {
    this._id = value;
  }

  get nom(): string {
    return this._nom;
  }

  set nom(value: string) {
    this._nom = value;
  }

  get nbPersonne(): number {
    return this._nbPersonne;
  }

  set nbPersonne(value: number) {
    this._nbPersonne = value;
  }

  get temps(): number {
    return this._temps;
  }

  set temps(value: number) {
    this._temps = value;
  }

  get cout(): Cout {
    return this._cout;
  }

  set cout(value: Cout) {
    this._cout = value;
  }

  get tags(): string[] {
    return this._tags;
  }

  set tags(value: string[]) {
    this._tags = value;
  }

  get ingredients(): Ingredient[] {
    return this._ingredients;
  }

  set ingredients(value: Ingredient[]) {
    this._ingredients = value;
  }

  get etapes(): EtapeRecette[] {
    return this._etapes;
  }

  set etapes(value: EtapeRecette[]) {
    this._etapes = value;
  }

  get picture(): ImageModel {
    return this._picture;
  }

  set picture(value: ImageModel) {
    this._picture = value;
  }

  get auteur(): Utilisateur {
    return this._auteur;
  }

  set auteur(value: Utilisateur) {
    this._auteur = value;
  }

  get difficulte(): Difficulte {
    return this._difficulte;
  }

  set difficulte(value: Difficulte) {
    this._difficulte = value;
  }

  get dateCreation(): Date {
    return this._dateCreation;
  }

  set dateCreation(value: Date) {
    this._dateCreation = value;
  }

  get commentaires(): Commentaire[] {
    return this._commentaires;
  }

  set commentaires(value: Commentaire[]) {
    this._commentaires = value;
  }
  get type(): string {
    return this._type;
  }

  set type(value: string) {
    this._type = value;
  }
}
